Lib/warnings.py

```diff
@@ -488,6 +488,23 @@ def __exit__(self, *exc_info):
         self._module._showwarnmsg_impl = self._showwarnmsg_impl
 
 
+# Private utility function called by _PyErr_WarnUnawaitedCoroutine
+def _warn_unawaited_coroutine(coro):
+    msg_lines = [
+        f"coroutine '{coro.__qualname__}' was never awaited\n"
+    ]
+    if coro.cr_origin is not None:
+        import linecache, traceback
+        def extract():
+            for filename, lineno, funcname in reversed(coro.cr_origin):
+                line = linecache.getline(filename, lineno)
+                yield (filename, lineno, funcname, line)
+        msg_lines.append("Coroutine created at (most recent call last)\n")
+        msg_lines += traceback.format_list(list(extract()))
+    msg = "".join(msg_lines).rstrip("\n")
+    warn(msg, category=RuntimeWarning, stacklevel=2, source=coro)
```

Comment:
Small, easy to miss point here: passing `source=coro` here means that if the user happens to have `tracemalloc` enabled, its origin traceback will be included in the warning. Kinda neat! It does mean if you have *both* `set_coroutine_origin_tracking_depth` and `tracemalloc` enabled you'll get two at-least-partially-redundant tracebacks in the same warning. I don't know that I care that much about this but FYI.

Reply:
I think the redundant warnings are OK. `tracemalloc` is enabled quite rarely.  I also think what you wrote in this comment should be reflected in a code comment in case someone debugs the double warnings issue :)